src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+    switch (this->currentStep)
+    {
+        case stepBeginOfRecord:
+            goto stepBeginOfRecord;
+
+        case stepReadHeader:
+            goto stepReadHeader;
+
+        case stepReadBody:
+            goto stepReadBody;
+    }
+
+    if (this->page_offset == 0 || this->page_offset >= this->walPageSize)
+    {
+stepBeginOfRecord:
+        if (readPage(this, input, stepBeginOfRecord))
+        {
+            return false;
+        }
+
+        if (this->is_begin)
+        {
+            this->is_begin = false;
+            // There may be an unfinished record from the previous file at the beginning of the file. Just skip it.
+            // We should have handled it elsewhere.
+            if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD &&
+                !(this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD))
+            {
+                this->page_offset += MAXALIGN(this->current_header->xlp_rem_len);
+            }
+        }
+        else if (this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD)
+        {
+            THROW_FMT(FormatError, "should not be XLP_FIRST_IS_CONTRECORD");
+        }
+    }
+
+    // Record header can be split between pages but first field xl_tot_len is always on single page
+    uint32_t record_size = getRecordSize(this->page + this->page_offset);
+
+    if (this->rec_buf_size < record_size)
+    {
+        MEM_CONTEXT_OBJ_BEGIN(this)
+        {
+            this->record = memResize(this->record, record_size);
+        }
+        MEM_CONTEXT_OBJ_END();
+        this->rec_buf_size = record_size;
+    }
+
+    memcpy(this->record, this->page + this->page_offset, Min(SizeOfXLogRecord, this->walPageSize - this->page_offset));
+
+    this->tot_len = this->record->xl_tot_len;
+
+    // If header is split read rest of the header from next page
+    if (SizeOfXLogRecord > this->walPageSize - this->page_offset)
+    {
+        this->got_len = this->walPageSize - this->page_offset;
+stepReadHeader:
+        if (readPage(this, input, stepReadHeader))
+        {
+            return false;
+        }
+
+        if (this->current_header->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return true;
+        }
+
+        if (!(this->current_header->xlp_info & XLP_FIRST_IS_CONTRECORD))
+        {
+            THROW_FMT(FormatError, "should be XLP_FIRST_IS_CONTRECORD");
+        }
+
+        memcpy(((char *) this->record) + this->got_len, this->page + this->page_offset,
+               SizeOfXLogRecord - this->got_len);
+        this->tot_len -= this->got_len;
+        this->header_offset = SizeOfXLogRecord - this->got_len;
+    }
+    else
+    {
+        this->header_offset = SizeOfXLogRecord;
+    }
+    this->got_len = SizeOfXLogRecord;
```

Comment:
How about calculating `got_len` and then making sure it is equal to `SizeOfXLogRecord`?

Reply:
I can't think of a case where we couldn't read the header completely except for the exhaustion of the input buffer, which is already being handled.